code/modules/particles/byond_particles/particool.dm

```diff
@@ -0,0 +1,253 @@
+			scale	vector (2D)	Scale applied to icon, if used; defaults to list(1,1)
+			grow	num	Change in scale per tick; defaults to list(0,0)
+			rotation	num	Angle of rotation (clockwise); applies only if using an icon
+			spin	num	Change in rotation per tick
+			friction	num	Amount of velocity to shed (0 to 1) per tick, also applied to acceleration from drift
+		Vars that are evalulated every tick
+			drift	vector	Added acceleration every tick; e.g. a circle or sphere generator can be applied to produce snow or ember effects
+*/
+
+//This should should have a default at some point
+GLOBAL_LIST_INIT(master_particle_info, list())
+
+/// Takes an input string like "A" and turns it into a macro A
+#define TEXT_TO_MACRO(Y) if(#Y) return Y
+
+//some helpyers
+/datum/particle_editor/proc/ListToMatrix(list/L)
+
+	//Normal Matrixes allow 6
+	switch(length(L))
+		if(6)
+			return matrix(L[1],L[2],L[3],L[4],L[5],L[6])
+
+/datum/particle_editor/proc/stringToList(str, toNum = FALSE, restoreNull = FALSE)
+	. = splittext(str,regex(@"(?<!\\),"))
+	if(toNum)
+		for(var/i = 1; i <= length(.); ++i)
+			.[i] = stringToNum(.[i], restoreNull)
+	if(length(.) == 1 && (isnull(.[1]) || .[1] == ""))
+		. = null
+
+/datum/particle_editor/proc/stringToNum(str, restoreNull = FALSE)
+	. = text2num(str)
+	if(isnull(.) && restoreNull)
+		. = str
+
+/datum/particle_editor/proc/stringToMatrix(str)
+	return ListToMatrix(stringToList(str))
+
+
+
+
+/datum/particle_editor
+	var/atom/movable/target
+
+/datum/particle_editor/New(atom/target)
+	src.target = target
+
+/datum/particle_editor/ui_state(mob/user)
+	return GLOB.admin_state
+
+/datum/particle_editor/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Particool")
+		ui.open()
+
+/datum/particle_editor/ui_static_data(mob/user)
+	var/list/data = list()
+	data["particle_info"] = GLOB.master_particle_info
+	return data
+
+/datum/particle_editor/ui_data()
+	var/list/data = list()
+	data["target_name"] = target.name
+	data["target_particle"] = getParticleVars()
+	return data
+
+/datum/particle_editor/proc/getParticleVars()
+	. = target.particles ? target.particles.vars : null
+
+
+//expects an assoc list of name, type, value - type must be in this list
+/datum/particle_editor/proc/translate_value(list/L)
+	//string/float/int come in as the correct type, so just whack em in
+
+	switch(L["type"])
+		if("string") return L["value"]
+		if("float") return L["value"]
+		if("int") return L["value"]
+		if("vector") return L["value"]
+		if("vector2") return L["value"]
+		if("color") return stringToNum(L["value"], TRUE)
+		if("text") return L["value"]
+		if("list") return stringToList(L["value"], TRUE, TRUE)
+		if("numList") return stringToList(L["value"],TRUE)
+		if("matrix") return ListToMatrix(L["value"])
+		if("generator") return generateGenerator(L["value"]) // This value should be a new list, if it isn't then we will explode
+
+/datum/particle_editor/proc/generateGenerator(L)
```

Comment:
Looks like L is being treated as a list here, shouldn't this be ``list/L`` in args?